pkg/action/action_handler.go

```diff
@@ -1,153 +1,208 @@
+
+	glog.V(2).Infof("Action %s for %s-%s succeeded.", content.ActionType, content.TargetObject.TargetObjectType, content.TargetObject.TargetObjectName)
+	progress := int32(100)
+	h.sendActionResult(proto.ActionResponseState_SUCCEEDED, progress, "Success")
 }
 
-func (this *ActionHandler) getNextFailedVMTEvent() {
-	e, err := this.config.FailedVMTEventQueue.Pop(nil)
-	if err != nil {
-		//TODO
-	}
-	event := e.(*registry.VMTEvent)
+func (h *ActionHandler) getNextFailedTurboAction() {
+	event := <-h.failedActionChan
 
 	glog.V(3).Infof("Failed event is %v", event)
-	// TODO, send back action failed
 	content := event.Content
-	msgID := int32(content.VMTMessageID)
-	if msgID > -1 {
-		glog.V(2).Infof("Action %s for %s failed.", content.ActionType, content.TargetSE)
-		progress := int32(0)
-		this.sendActionResult(proto.ActionResponseState_FAILED, progress, msgID, "Failed")
-	}
-	// TODO init discovery
+
+	glog.V(2).Infof("Action %s for %s-%s failed.", content.ActionType, content.TargetObject.TargetObjectType, content.TargetObject.TargetObjectName)
+	progress := int32(0)
+	h.sendActionResult(proto.ActionResponseState_FAILED, progress, "Failed 1")
+}
+
+// Implement ActionExecutorClient interface defined in Go SDK.
+// Execute the current action and return the action result.
+func (h *ActionHandler) ExecuteAction(actionExecutionDTO *proto.ActionExecutionDTO,
+	accountValues []*proto.AccountValue,
+	progressTracker sdkprobe.ActionProgressTracker) (*proto.ActionResult, error) {
+
+	actionItems := actionExecutionDTO.GetActionItem()
+	// TODO: only deal with one action item.
+	actionItemDTO := actionItems[0]
+	h.execute(actionItemDTO)
+
+	result := <-h.resultChan
+	glog.V(4).Infof("Action result is %++v", result)
+	// TODO: currently the code in SDK make it share the actionExecution client between different workers. Once it is changed, need to close the channel.
+	//close(h.config.StopEverything)
+	return result, nil
 }
 
-func (this *ActionHandler) Execute(actionItem *proto.ActionItemDTO, msgID int32) {
-	_, err := this.actionExecutor.ExcuteAction(actionItem, msgID)
+func (h *ActionHandler) execute(actionItem *proto.ActionItemDTO) {
+	actionType, err := getActionTypeFromActionItemDTO(actionItem)
 	if err != nil {
-		glog.Errorf("Error execute action: %s", err)
-		this.sendActionResult(proto.ActionResponseState_FAILED, int32(0), msgID, "Failed")
+		glog.Errorf("Failed to execute action: %s", err)
+		h.sendActionResult(proto.ActionResponseState_FAILED, int32(0), "Failed to execute action")
+		return
+	}
+	executor, exist := h.actionExecutors[actionType]
+	if !exist {
+		glog.Errorf("action type %s is not support", actionType)
+		h.sendActionResult(proto.ActionResponseState_FAILED, int32(0), "Failed to execute action")
+		return
 	}
+
+	action, err := executor.Execute(actionItem)
+	if err != nil {
+		glog.Errorf("Failed to execute action: %s", err)
+		h.sendActionResult(proto.ActionResponseState_FAILED, int32(0), "Failed to execute action")
```

Comment:
Yes. But I think you are not talking about simply print out the error returned, right?